### ISSUE TITLE
FR: allow refresh flags to be changed at runtime

### ISSUE THREAD

**Issue by aconicq** [koreader/koreader#1375]
Night mode refreshes every page rather than using the custom refresh page number. Normal (day) mode works as intended.

**Comment by ec-cwang** [comment]
Mine has the same problem (kpw2 5.4.5.1, koreader version nightly 20141207), but the night mode refresh works fine in the version nightly 20141031.

**Comment by chrox** [comment]
@ec-cwang Could you also try to locate at which point the night mode breaks? There are [more than 250 commits](https://github.com/koreader/koreader/compare/v2014.10.31-nightly...v2014.12.07-nightly) between build 20141207 and 20141031.

**Comment by ec-cwang** [comment]
@chrox Sorry for being not accurate, I just did a test of the builds between 20141031 and 20141207 and I found that the "night mode is forced to refresh on every page " begins at build 20141110 (koreader-kindle-arm-linux-gnueabi-v2014.04-1008-g779fc5a). Both of build 20141031 and 20141107 work fine.

**Comment by chrox** [comment]
Thanks, that's better now. We can narrow down the suspicious commits [here](https://github.com/koreader/koreader/compare/v2014.11.07-nightly...v2014.11.10-nightly). So pinging @NiLuJe.

**Comment by hwhw** [comment]
I did a major rework of what @NiLuJe implemented back then. But I left all the device-dependent refresh modes in place. Before that, we used sub-optimal modes - for normal mode. I never implemented separate modes for night mode.

So it boils down to:
* we switched to a refresh mode better fit for *normal* display. We'll likely want to keep that.
* since there is no separate mode config for night mode, that switched, too.

What we need to do:
* have separate modes for night mode
* find the right refresh modes for night mode and configure them

**Comment by hwhw** [comment]
This is where we need to start:
https://github.com/koreader/koreader-base/blob/master/ffi/framebuffer.lua#L290

And the device specifics should go into framebuffer_....lua, with a general abstraction in framebuffer.lua.

**Comment by bel3atar** [comment]
I have the same problem with epub(pdf is fine) be it in nightmode or not.
Latest stable and latest nightly are both affected.
PW2

**Comment by NiLuJe** [comment]
Don't remember if we ended up having nightmode-specific waveforme updates, but if we don't, we'd need to to handle this issue on reagl devices.

FW >= 5.5 (theoretically) has a reagl-aware nightmode-friendly refresh mode, but that'd mean an extra check for underlying FW version somewhere.

And since I thoroughlyy don't care about nightmode, patches welcome? ;).

**Comment by houqp** [comment]
Can someone with reagl devices help check out refresh mode (GL16_INV) and the HW inversion solutions? Looks like we only need to set an attribute in `toggleNightMode` function.

**Comment by NiLuJe** [comment]
I do have a PW2 on 5.7.2, so I might be able to try some things at one point.

**Comment by roger64** [comment]
I confirm that on my PW3 (5.6.5), I receive a flash at every page, Koreader does not apply the six page refresh setting. I just close the eyes before turning  ;)

**Comment by NiLuJe** [comment]
Quick'n dirty patch to check how GL16_INC & HW inversion work:

```
diff --git a/ffi/framebuffer_mxcfb.lua b/ffi/framebuffer_mxcfb.lua
index 029f4cb..b2f8d31 100644
--- a/ffi/framebuffer_mxcfb.lua
+++ b/ffi/framebuffer_mxcfb.lua
@@ -235,7 +235,8 @@ local function refresh_k51(fb, refreshtype, waveform_mode, wait, x, y, w, h)
        refarea[0].temp = ffi.C.TEMP_USE_AUTO
        -- NOTE: We never use any flags on Kindle.
        -- TODO: EPDC_FLAG_ENABLE_INVERSION & EPDC_FLAG_FORCE_MONOCHROME might be of use, though...
-       refarea[0].flags = 0
+       --refarea[0].flags = 0
+       refarea[0].flags = ffi.C.EPDC_FLAG_ENABLE_INVERSION
 
        return mxc_update(fb, refarea, refreshtype, waveform_mode, wait, x, y, w, h)
 end
@@ -321,8 +322,9 @@ function framebuffer:init()
             self.wait_for_marker_ui = false
         elseif self.device.model == "KindlePaperWhite2" then
             self.mech_wait_update_complete = kindle_carta_mxc_wait_for_update_complete
-            self.waveform_partial = ffi.C.WAVEFORM_MODE_REAGL
-            self.update_mode_partial = ffi.C.UPDATE_MODE_FULL -- REAGL get upgraded to full
+            --self.waveform_partial = ffi.C.WAVEFORM_MODE_REAGL
+            --self.update_mode_partial = ffi.C.UPDATE_MODE_FULL -- REAGL get upgraded to full
+            self.waveform_partial = ffi.C.WAVEFORM_MODE_GL16_INV
             self.wait_for_marker_full = true
             self.wait_for_marker_partial = true
             self.wait_for_marker_fast = true
```

A very quick test on my PW2 on 5.7.2 shows that it works.
It's still ghosty as hell, but probably less so than usual. (Then again, I'm not the right guy to judge that, I hate nightmode with a fiery passion ;)).

Currently keeps the intermittent full flash as GC16, which I think is the right thing to do, since GL* modes are designed to be PARTIAL only.

Also, I lied, it's not a REAGL mode at all, I was mixing up stuff ;).

**Comment by Frenzie** [comment]
I don't like night mode either, so I'll refrain from judgment. :P

**Comment by Markismus** [comment]
I rather like night mode as a way to limit light pollution for the woman sleeping next to me. So I'll look into it tomorrow. An improvement in night mode is indeed needed!

I can't remember clearly, but wasn't REAGL disabled in normal mode by the koreader code because it results were sketchy at best?

**Comment by Markismus** [comment]
Just found #741. Does this also apply to the Aura H2O?

**Comment by NiLuJe** [comment]
@Markismus: No, it was disabled on devices where it's not supported by the hardware. The original implementation was enabled on slightly too many devices, because I wrongly assumed Kobo successfully kept using it on the devices succeeding Aura, whereras in fact they backtracked and forgot all about it, for some unfathomable reason.

**Comment by Markismus** [comment]
@NiLuJe So has the Aura H2O REAGL or not? I thought it had, but it is a successor of the Aura, so it hasn't?

**Comment by Markismus** [comment]
Odd. Because I do see that the screen generated by Nickel is different from Koreader's. I always assumed it was due to using REAGL.

**Comment by NiLuJe** [comment]
@Markismus: No REAGL on the H2O (as well as on the Glo HD, AFAIK), unfortunately.

What do you mean by your latest comment (and/or how do you measure it)?

Last I checked, we're sending roughly the exact same ioctls as Nickel, AFAICT through strace (which proved to be accurate on the Kindle, where we have extra native debug features to double-check that).

**Comment by Markismus** [comment]
No measureable, but the page on Nickel seem more clear. As if it's on brighter paper with a sharper font, when you compare the same epub. Could be the font of course, but it seems more to be something about the contrast. Odd, it being the same screen and all. 

The vagueness about these observations are enough not to bother with Nickel for long, but it would imply that there is still room for improvement in koreader's handling of epubs. Would have been nice if it was down to REAGL, because I sure don't have a clue what else could have an impact on the contrast.

**Comment by NiLuJe** [comment]
@Markismus: You can try switching to the native hinter if that's not already the case (cf #706).

**Comment by Markismus** [comment]
I've changed font.hinting.mode from 2 to 1 (BCI). It's seems sharper, but I'll have to change it back to see whether it's not due to my wishful thinking.

**Comment by houqp** [comment]
@Markismus so do you have this issue with your H2O too?

**Comment by NiLuJe** [comment]
Not AFAIK ;).

**Comment by chrox** [comment]
No progress, please reopen this if you still have problem with the latest nightly build.

**Comment by NiLuJe** [comment]
(Left it closed, it's just a "nice to have" thing if I ever get the time ;)).

**Comment by dsanson** [comment]
I am still having the problem with the latest nightly build on my PW2 (running 5.6.5). @NiLuJe's patch solves the problem, though now night mode and regular mode seem to be swapped.

**Comment by NiLuJe** [comment]
@dsanson: Yep, that was just a POC.

I do intend to try to work on that someday, but to be frank, it's quite low on the pile of things to do ;).

**Comment by dsanson** [comment]
Sad it is a low priority! Wish I understood the code better so I could work on it.

The bright flash with each page turn makes night mode almost unusable, especially for, as @Markismus put it, the "woman sleeping next to me". For me at least, your proof of concept patch is better than nothing.

**Comment by aconicq** [comment]
I'd like to reopen this issue. Night mode page refreshing occurs every page regardless of settings. I love using night mode with long durations between refreshes during bedtime reading. This virtually eliminates all unnecessary light and I don't find the ghosting overwhelming. 

Is anyone able to look into this issue?
Many thanks

**Comment by houqp** [comment]
If @NiLuJe 's patch actually worked, then it seems like a straight forward fix for me. We just need to pass in different flags based on the mode setting. If anyone want to take a stab at it, please let me know, I can help point you to the right direction.

**Comment by bel3atar** [comment]
Also noticed that when night mode is disabled, some PDF do refresh every page while others don't.

**Comment by aconicq** [comment]
I am keen to fix this. I can also confirm that NiLuJe's patch works, and if only the second part is implemented then the night/day mode is in the correct order. It seems to me that this solution is fine and should be implemented into the nightly builds. What do you mean @houqp about different flags?

**Comment by houqp** [comment]
@aconicq awesome! It's late here now, so I can't type a long answer, please let me know if what I said does not make sense to you.

If you look at the patch, it changed couple default flags for screen fresh action, namely `refarea[0].flags`, `self.waveform_partial` and `self.update_mode_partial`:

```
+       --refarea[0].flags = 0
+       refarea[0].flags = ffi.C.EPDC_FLAG_ENABLE_INVERSION

-            self.waveform_partial = ffi.C.WAVEFORM_MODE_REAGL
-            self.update_mode_partial = ffi.C.UPDATE_MODE_FULL -- REAGL get upgraded to full
+            --self.waveform_partial = ffi.C.WAVEFORM_MODE_REAGL
+            --self.update_mode_partial = ffi.C.UPDATE_MODE_FULL -- REAGL get upgraded to full
+            self.waveform_partial = ffi.C.WAVEFORM_MODE_GL16_INV
```

Now all we have to do is to make these two section of the code set those flags based on attributes from `self` instead of hardcoded values. `self` here being the `framebuffer` in `ffi/framebuffer_mxcfb.lua`. Then we add an method to framebuffer to set those flags attributes dynamically when Night mode is toggled on or off. In the reader frontend code, `Screen` is the `framebuffer`.

**Comment by NiLuJe** [comment]
The issue is that some of this is FW and/or device specific, IIRC.

i.e., the new waveform modes are FW 5.5 and up, and the hardware inversion
flag is broken and/or crashy on (some?) kobos...

On Nov 16, 2016 10:22 AM, "Qingping Hou" <notifications@github.com> wrote:

@aconicq <https://github.com/aconicq> awesome! It's late here now, so I
can't type a long answer, please let me know if what I said does not make
sense to you.

If you look at the patch, it changed couple default flags for screen fresh
action, namely refarea[0].flags, self.waveform_partial and
self.update_mode_partial:

+       --refarea[0].flags = 0
+       refarea[0].flags = ffi.C.EPDC_FLAG_ENABLE_
INVERSION

-            self.waveform_partial = ffi.C.WAVEFORM_MODE_REAGL
-            self.update_mode_partial = ffi.C.UPDATE_MODE_FULL --
REAGL get upgraded to full
+            --self.waveform_partial = ffi.C.WAVEFORM_MODE_REAGL
+            --self.update_mode_partial = ffi.C.UPDATE_MODE_FULL --
REAGL get upgraded to full
+            self.waveform_partial = ffi.C.WAVEFORM_MODE_GL16_INV

Now all we have to do is to make these two section of the code set those
flags based on attributes from self instead of hardcoded values. self here
being the framebuffer in ffi/framebuffer_mxcfb.lua. Then we add an method
to framebuffer to set those flags attributes dynamically when Night mode is
toggled on or off. In the reader frontend code, Screen is the framebuffer.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/koreader/koreader/issues/1375#issuecomment-260895179>,
or mute the thread

**Comment by houqp** [comment]
> The issue is that some of this is FW and/or device specific, IIRC.
> 
> i.e., the new waveform modes are FW 5.5 and up, and the hardware inversion
> flag is broken and/or crashy on (some?) kobos... 

I see, now I understand why you said it's a dirty patch ;P

Regardless, I think we will need the ability to change these refresh flags at runtime. Once that feature is enabled, we can attack this problem in two steps:

* turn the refresh flags into a setting so each user can use whatever setting that works for his/her device
* as more user report what flags work for what models/FWs, we can do the auto model detection when reader starts up and set the correct flags accordingly

**Comment by NiLuJe** [comment]
I still have zero desire to work on this, but everything should now be in a sane state :).

**Comment by NiLuJe** [comment]
Random x-ref: #2571

And my latest [braindump](https://www.mobileread.com/forums/showpost.php?p=3747576&postcount=60) on the subject, if anyone wants to take a stab at this...

The most convoluted part would probably indeed involve handling images right in CRe, the rest is fairly approachable if one is determined enough.

**Comment by NiLuJe** [comment]
Fixed via #4871 ;).
